customer_outstanding_statement/report/customer_outstanding_statement.py

```diff
@@ -0,0 +1,322 @@
+# Copyright 2018 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from datetime import datetime, timedelta
+from odoo.tools.misc import DEFAULT_SERVER_DATE_FORMAT
+from odoo import api, fields, models
+
+
+class CustomerOutstandingStatement(models.AbstractModel):
+    """Model of Customer Outstanding Statement"""
+
+    _name = 'report.customer_outstanding_statement.statement'
+
+    def _format_date_to_partner_lang(self, str_date, partner_id):
+        lang_code = self.env['res.partner'].browse(partner_id).lang
+        lang = self.env['res.lang']._lang_get(lang_code)
+        date = datetime.strptime(str_date, DEFAULT_SERVER_DATE_FORMAT).date()
+        return date.strftime(lang.date_format)
+
+    def _display_lines_sql_q1(self, partners, date_end):
+        return """
+            SELECT m.name as move_id, l.partner_id, l.date, l.name,
+                            l.ref, l.blocked, l.currency_id, l.company_id,
+            CASE WHEN (l.currency_id is not null AND l.amount_currency > 0.0)
+                THEN sum(l.amount_currency)
+                ELSE sum(l.debit)
+            END as debit,
+            CASE WHEN (l.currency_id is not null AND l.amount_currency < 0.0)
+                THEN sum(l.amount_currency * (-1))
+                ELSE sum(l.credit)
+            END as credit,
+            CASE WHEN l.balance > 0.0
+                THEN l.balance - sum(coalesce(pd.amount, 0.0))
+                ELSE l.balance + sum(coalesce(pc.amount, 0.0))
+            END AS open_amount,
+            CASE WHEN l.balance > 0.0
+                THEN l.amount_currency - sum(coalesce(pd.amount_currency, 0.0))
+                ELSE l.amount_currency + sum(coalesce(pc.amount_currency, 0.0))
+            END AS open_amount_currency,
+            CASE WHEN l.date_maturity is null
+                THEN l.date
+                ELSE l.date_maturity
+            END as date_maturity
+            FROM account_move_line l
+            JOIN account_account_type at ON (at.id = l.user_type_id)
+            JOIN account_move m ON (l.move_id = m.id)
+            LEFT JOIN (SELECT pr.*
+                FROM account_partial_reconcile pr
+                INNER JOIN account_move_line l2
+                ON pr.credit_move_id = l2.id
+                WHERE l2.date <= '%s'
+            ) as pd ON pd.debit_move_id = l.id
+            LEFT JOIN (SELECT pr.*
+                FROM account_partial_reconcile pr
+                INNER JOIN account_move_line l2
+                ON pr.debit_move_id = l2.id
+                WHERE l2.date <= '%s'
+            ) as pc ON pc.credit_move_id = l.id
+            WHERE l.partner_id IN (%s) AND at.type = 'receivable'
+                                AND not l.reconciled AND l.date <= '%s'
```

Comment:
@mreficient but it's from here, we should filter at the end with open balance different than 0.

Reply:
Yes, yes, you are right.